### PR TITLE
Add ESP32SPI support

### PR DESCRIPTION
**Pull request by justmobilize**
Add support for ESP32SPI, which would allow this to be used with:
- Onboard WiFi
- WIZNET5k
- ESP32SPI

Merged code from: https://github.com/adafruit/Adafruit_CircuitPython_ESP32SPI/blob/main/examples/esp32spi_udp_client.py

### CODE REVIEW

**Comment by justmobilize** [comment]
@anecdata and @dhalbert - simple update

**Review by anecdata — APPROVED** [review]
This looks good to me! Tested on config with ESP32-S3 and ESP32SPI. Code looks fine, and the two components were each already reviewed when they were merged into their respective repos.

<details>
<summary>test code...</summary>

```py
# also uses esp32spi pr#199 to simplify connect and ipv4_address
import time
import os
import board
import digitalio
import wifi
import adafruit_connection_manager
from adafruit_esp32spi import adafruit_esp32spi
import adafruit_ntp

radios = []
pools = []
connection_managers = []
ntps = []

# native wifi
radios.append(wifi.radio)

# esp32spi
spi = board.SPI()
esp32_cs = digitalio.DigitalInOut(board.D13)
esp32_reset = digitalio.DigitalInOut(board.D12)
esp32_ready = digitalio.DigitalInOut(board.D11)
radios.append(adafruit_esp32spi.ESP_SPIcontrol(spi, esp32_cs, esp32_ready, esp32_reset))

for radio in range(0, len(radios)):
    pools.append(adafruit_connection_manager.get_radio_socketpool(radios[radio]))
    connection_managers.append(adafruit_connection_manager.get_connection_manager(pools[radio]))
    radios[radio].connect(os.getenv("WIFI_SSID"), os.getenv("WIFI_PASSWORD"))
    ntps.append(adafruit_ntp.NTP(pools[radio], tz_offset=0))

while True:
    for radio in range(0, len(radios)):
        radios[radio].connect(os.getenv("WIFI_SSID"), os.getenv("WIFI_PASSWORD"))
        print(f'{radios[radio].ipv4_address} {ntps[radio].datetime}')
        time.sleep(1)
```
</details>

**Comment by anecdata** [comment]
We may want to ad this to the list, like WIZnet, where the init wouldn't necessarily require a network connection.

File: adafruit_ntp.py [hunk 1]
```diff
@@ -53,9 +57,8 @@ def __init__(
         :param int socket_timeout: UDP socket timeout, in seconds.
         """
         self._pool = socketpool
-        self._server = server
-        self._port = port
-        self._packet = bytearray(48)
+        self._socket_address = self._pool.getaddrinfo(server, port)[0][4]
```
Comment (justmobilize) [inline adafruit_ntp.py:60]: @anecdata good call out on the `init`. I'll move this to the `datetime` property:
```
if self._socket_address is None:
   self._socket_address = self._pool.getaddrinfo(server, port)[0][4]
```
Will hold on other reviews first

**Comment by jepler** [comment]
It appears that the main change is to work around missing udp "sendto" support in wiznet. Did anyone investigate whether sendto (and any other missing functionality) can be added to wiznet instead?

**Comment by jepler** [comment]
It appears that arduino's version of nina-fw supports socket_sendto but adafruit's fork does not.

if send can substitute for sendto, is there any reason to *retain* the two runtime branches (sendto & connect+send) rather than just switching to connect+send with a note that it's used for compatibility with nina-fw / esp32spi?

**Comment by justmobilize** [comment]
@jepler I will try that. since sendto, basically just does that - it should work

**Comment by anecdata** [comment]
My impression has been that `connect` isn't typically used for UDP. In this case it's an artifact of the non-standard way Arduino (NINA) handles sockets. Virtually none of the typical socket code that runs on ÇPython, native wifi, or WIZnet, work in NINA.

**Comment by justmobilize** [comment]
@jepler the native socket doesn't have `recv`, so either way we are doing special code per chip. This seems the most straightforward... Thoughts?

**Comment by jepler** [comment]
as the return value of `recvfrom_into` is ignored, can't `recv_into` be substituted?

**Comment by justmobilize** [comment]
But the esp32spi only has recv...

**Comment by justmobilize** [comment]
@jepler anything thoughts / recommendations?

**Comment by jepler** [comment]
https://github.com/adafruit/Adafruit_CircuitPython_ESP32SPI/blob/main/adafruit_esp32spi/adafruit_esp32spi_socketpool.py#L137 I see "def recv_into", is it not applicable?

**Comment by justmobilize** [comment]
@jepler okay, I got it as close as I can. The only difference now is the connect method

**Comment by jepler** [comment]
Maybe [#201](https://github.com/adafruit/Adafruit_CircuitPython_ESP32SPI/pull/201) helps with the connect() difference

**Comment by justmobilize** [comment]
@jepler that works

**Comment by justmobilize** [comment]
@anecdata willing to re-test with the newest esp32spi?

**Review by jepler — APPROVED** [review]
thanks for your patience on this. I'm much happier now that nothing's conditional based on the socket implementation; I think this is the best we can do without changes to the nina-fw running on the airlift microcontrollers.

**Comment by justmobilize** [comment]
@jepler thanks for pushing. I also like this better!

**Review by anecdata — COMMENTED** [review]
Looks good! Using _adafruit_esp32spi_ `socket.connect(): Auto-select UDP_MODE #201`, this NTP PR is repeatedly successful on native wifi, esp32spi, and WIZnet, with no changes to user code or library needed. NTP sessions from each device IP address confirmed at the router.

<details>
<summary>test code...</summary>

```py
import time
import os
import traceback
import board
import digitalio
import wifi
import adafruit_connection_manager
from adafruit_esp32spi import adafruit_esp32spi
from adafruit_wiznet5k import adafruit_wiznet5k
import adafruit_ntp
import adafruit_requests


URL = "http://wifitest.adafruit.com/testwifi/index.html"

# General
def setup(radio):
    radios.append(radio)
    pool = adafruit_connection_manager.get_radio_socketpool(radio)
    ntp = adafruit_ntp.NTP(pool, tz_offset=0)
    ntps.append(ntp)

def connect(radio):
    if radio.__class__.__name__ == "Radio":
        ipv4_str = connect_native(radio)
    elif radio.__class__.__name__ == "WIZNET5K":
        ipv4_str = connect_eth(radio)
    elif radio.__class__.__name__ == "ESP_SPIcontrol":
        ipv4_str = connect_esp(radio)
    return ipv4_str

# Native wifi
def setup_native():
    radio = wifi.radio
    setup(radio)

def connect_native(radio):
    while not radio.connected:
        try:
            radio.connect(os.getenv("WIFI_SSID"), os.getenv("WIFI_PASSWORD"))
        except Exception as ex:
            traceback.print_exception(ex, ex, ex.__traceback__)
    ipv4_str = radio.ipv4_address
    return ipv4_str

# WIZnet
def setup_eth():
    cs = digitalio.DigitalInOut(board.A3)
    mac = f'de:ad:be:ef:fe:{252:02x}'  # need this to avoid DHCP collisions
    radio = adafruit_wiznet5k.WIZNET5K(spi, cs, mac=mac, debug=False)
    setup(radio)

def connect_eth(radio):
    ipv4_str = radio.pretty_ip(radio.ip_address)
    return ipv4_str

# ESP32SPI
def setup_esp():
    esp32_cs = digitalio.DigitalInOut(board.A0)
    esp32_reset = digitalio.DigitalInOut(board.A1)
    esp32_ready = digitalio.DigitalInOut(board.A2)
    radio = adafruit_esp32spi.ESP_SPIcontrol(spi, esp32_cs, esp32_ready, esp32_reset, debug=False)
    setup(radio)

def connect_esp(radio):
    while not radio.is_connected:
        try:
            radio.connect_AP(os.getenv("WIFI_SSID"), os.getenv("WIFI_PASSWORD"))
        except Exception as ex:
            traceback.print_exception(ex, ex, ex.__traceback__)
    ipv4_str = radio.pretty_ip(radio.ip_address)
    return ipv4_str


time.sleep(3)  # wait for serial

spi = board.SPI()

radios = []
ntps = []
setup_native()
setup_eth()
setup_esp()
while True:
    print(f'{"-"*25}')
    for radio in radios:
        print(connect(radio))
    for ntp in ntps:
        try:
            print(f'{ntp.datetime}')
            time.sleep(1)
        except Exception as ex:
            traceback.print_exception(ex, ex, ex.__traceback__)
```
</details>

**Comment by justmobilize** [comment]
@jepler (or maybe @dhalbert), can this be merged?

**Pull request merged**